core/commands/pin.go

```diff
@@ -61,22 +68,84 @@ var addPinCmd = &cmds.Command{
 			res.SetError(err, cmds.ErrNormal)
 			return
 		}
+		showProgress, _, _ := req.Option("progress").Bool()
 
-		added, err := corerepo.Pin(n, req.Context(), req.Arguments(), recursive)
-		if err != nil {
-			res.SetError(err, cmds.ErrNormal)
+		if !showProgress {
+			added, err := corerepo.Pin(n, req.Context(), req.Arguments(), recursive)
+			if err != nil {
+				res.SetError(err, cmds.ErrNormal)
+				return
+			}
+			res.SetOutput(&AddPinOutput{Pins: added})
 			return
 		}
 
-		res.SetOutput(&PinOutput{added})
+		v := new(dag.ProgressTracker)
+		ctx := context.WithValue(req.Context(), "progress", v)
+
+		ch := make(chan []*cid.Cid)
+		go func() {
+			defer close(ch)
+			added, err := corerepo.Pin(n, ctx, req.Arguments(), recursive)
+			if err != nil {
+				res.SetError(err, cmds.ErrNormal)
+				return
+			}
+			ch <- added
+		}()
+		out := make(chan interface{})
+		res.SetOutput((<-chan interface{})(out))
+		go func() {
+			defer close(out)
+			for {
+				timer := time.NewTimer(500 * time.Millisecond)
+				defer timer.Stop()
+				select {
+				case val, ok := <-ch:
+					if !ok {
+						// error already set just return
+						return
+					}
+					out <- &AddPinOutput{Progress: -1, Pins: val}
+					return
+				case <-timer.C:
+					out <- &AddPinOutput{Progress: v.Value()}
+				case <-ctx.Done():
+					res.SetError(ctx.Err(), cmds.ErrNormal)
+					return
+				}
+			}
+		}()
 	},
 	Marshalers: cmds.MarshalerMap{
 		cmds.Text: func(res cmds.Response) (io.Reader, error) {
-			added, ok := res.Output().(*PinOutput)
-			if !ok {
-				return nil, u.ErrCast()
-			}
+			var added []*cid.Cid
 
+			r, ok := res.Output().(*AddPinOutput)
+			if ok {
+				added = r.Pins
+			} else {
+				ch, ok := res.Output().(<-chan interface{})
+				if !ok {
+					return nil, u.ErrCast()
+				}
+				progressLine := false
+				for r0 := range ch {
+					r := r0.(*AddPinOutput)
+					if r.Progress == -1 || r.Pins != nil {
+						added = r.Pins
+					} else {
+						if progressLine {
```

Comment:
I use `\r` to replace write over the current line with a new progress line, but I only want to do that if the last line to stderr was a progress line